File: 0841-keys-and-rooms/0841-keys-and-rooms.cpp
class Solution {
public:
    void dfs(vector<int>&vis,int root,vector<vector<int>>& rooms,int n){
        vis[root] = 1;
        for(auto node : rooms[root]){
            if(!vis[node]){
                dfs(vis,node,rooms,n);
            }
        }
        return;
    }
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        int n = rooms.size();
        vector<int>vis(n,0);
        dfs(vis,0,rooms,n);
        for(auto it :vis){
            if(it==false){
                return false;
            }
        }
        return true;
    }
};

File: 0841-keys-and-rooms/0841-keys-and-rooms.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        
        def dfs(vis : set(),root: int,rooms: List[List[int]]):
            vis.add(root)
            for node in rooms[root]:
                if node not in vis:
                    dfs(vis,node,rooms)
            return
        
        n = len(rooms)
        vis = set()
        dfs(vis,0,rooms)
        if len(vis) == len(rooms):
            return True
        return False
    